packages/engine-server/src/markdown/remark/blockAnchors.ts

```diff
@@ -0,0 +1,90 @@
+import _ from "lodash";
+import { Eat } from "remark-parse";
+import Unified, { Plugin } from "unified";
+import { BlockAnchor, DendronASTDest } from "../types";
+import { MDUtilsV4 } from "../utils";
+
+// Letters, digits, dashes, and underscores.
+// The underscores are an extension over Obsidian.
+export const LINK_REGEX = /^\^([\w-]+)$/;
+export const LINK_REGEX_LOOSE = /\^([\w-]+)/;
+
+/**
+ *
+ * @param text The text to check if it matches an block anchor.
+ * @param matchLoose If true, a block anchor anywhere in the string will match. Otherwise the string must contain only the anchor.
+ * @returns The identifier for the match block anchor, or undefined if it did not match.
+ */
+export const matchBlockAnchor = (
+  text: string,
+  matchLoose: boolean = true
+): string | undefined => {
+  const match = (matchLoose ? LINK_REGEX_LOOSE : LINK_REGEX).exec(text);
+  if (match && match.length == 1) return match[1];
+  return undefined;
+};
+
+type PluginOpts = {};
+
+type CompilerOpts = {};
+
+const plugin: Plugin<[CompilerOpts?]> = function (
+  this: Unified.Processor,
+  opts?: PluginOpts
+) {
+  attachParser(this);
+  if (this.Compiler != null) {
+    attachCompiler(this, opts);
+  }
+};
+
+function attachParser(proc: Unified.Processor) {
+  function locator(value: string, fromIndex: number) {
+    return value.indexOf("^", fromIndex);
+  }
+
+  function inlineTokenizer(eat: Eat, value: string) {
+    const match = LINK_REGEX.exec(value);
+    if (match) {
+      return eat(match[0])({
+        type: "blockAnchor",
+        value,
+        id: match[1],
+      });
+    }
+    return;
+  }
+  inlineTokenizer.locator = locator;
+
+  const Parser = proc.Parser;
+  const inlineTokenizers = Parser.prototype.inlineTokenizers;
+  const inlineMethods = Parser.prototype.inlineMethods;
+  inlineTokenizers.blockAnchor = inlineTokenizer;
+  inlineMethods.splice(inlineMethods.indexOf("link"), 0, "blockAnchor");
+}
+
+function attachCompiler(proc: Unified.Processor, _opts?: CompilerOpts) {
+  const Compiler = proc.Compiler;
+  const visitors = Compiler.prototype.visitors;
+
+  if (visitors) {
+    visitors.blockAnchor = function (node: BlockAnchor) {
+      const { dest } = MDUtilsV4.getDendronData(proc);
+      switch (dest) {
+        case DendronASTDest.MD_DENDRON:
+        case DendronASTDest.MD_REGULAR:
+        case DendronASTDest.MD_ENHANCED_PREVIEW:
+        case DendronASTDest.HTML: {
+          // Anything more to do here? Can we embed HTML into the preview?
```

Comment:
yep, HTML can be embedded in the preview. that would be how I would implement blocks